updater/updater.go

```diff
@@ -0,0 +1,345 @@
+			Result: omaha.EventResultSuccess,
+		}
+	case ProgressError:
+		return &omaha.EventRequest{
+			Type:   omaha.EventTypeUpdateComplete,
+			Result: omaha.EventResultError,
+		}
+	default:
+		return nil
+	}
+}
+
+// OmahaRequestHandler wraps the Handle function which
+// takes in context, url, omaha.Request and
+// returns omaha.Response.
+type OmahaRequestHandler interface {
+	Handle(ctx context.Context, url string, req *omaha.Request) (*omaha.Response, error)
+}
+
+// Updater interface wraps functions required to update an application
+// CheckForUpdates checks if there are new updates version from
+// omaha server, ReportProgress reports progress of update
+// to the omaha server, ReportError reports errors with custom
+// error code, SendOmahaEvent,SendOmahaRequest sends omaha event request
+// and omaha request to omaha server respectively and TryUpdate function
+// takes an implementation of UpdateHandler and runs the complete flow
+// from checking updates to reporting successful installation. The
+// InstanceVersion and SetInstanceVersion functions are used to fetch
+// and set the current instance version of the updater.
+type Updater interface {
+	SendOmahaRequest(ctx context.Context, req *omaha.Request) (*omaha.Response, error)
+	SendOmahaEvent(ctx context.Context, event *omaha.EventRequest) (*omaha.Response, error)
+	CheckForUpdates(ctx context.Context) (*UpdateInfo, error)
+	ReportProgress(ctx context.Context, progress progress) error
+	ReportError(ctx context.Context, errorCode *int) error
+	CompleteUpdate(ctx context.Context, info *UpdateInfo) error
+	TryUpdate(ctx context.Context, handler UpdateHandler) error
+	InstanceVersion() string
+	SetInstanceVersion(version string)
+}
+
+// updater implements the Updater interface.
+type updater struct {
+	omahaURL      string
+	clientVersion string
+
+	instanceID      string
+	instanceVersion string
+	sessionID       string
+
+	appID   string
+	channel string
+
+	debug           bool
+	omahaReqHandler OmahaRequestHandler
+
+	mu sync.RWMutex
+}
+
+// Config is used to configure new updater instance.
+type Config struct {
+	OmahaURL        string
+	AppID           string
+	Channel         string
+	InstanceID      string
+	InstanceVersion string
+	Debug           bool
+	OmahaReqHandler OmahaRequestHandler
+}
+
+// New takes config and returns Updater and error,
+// returns an error if OmahaURL in the config is invalid.
+func New(config Config) (Updater, error) {
+	if _, err := url.Parse(config.OmahaURL); err != nil {
+		return nil, fmt.Errorf("parsing URL %q: %w", config.OmahaURL, err)
+	}
+
+	updater := updater{
+		omahaURL:        config.OmahaURL,
+		clientVersion:   defaultClientVersion,
+		instanceID:      config.InstanceID,
+		sessionID:       uuid.New().String(),
+		appID:           config.AppID,
+		instanceVersion: config.InstanceVersion,
+		channel:         config.Channel,
+		debug:           config.Debug,
+	}
+	updater.omahaReqHandler = config.OmahaReqHandler
```

Comment:
I guess that should be fine?
```suggestion
		debug:           config.Debug,
		omahaReqHandler: config.OmahaReqHandler
	}
```